python/ray/serve/_private/replica.py

```diff
@@ -1016,29 +1088,29 @@ async def call_user_method(
 
         return result
 
+    @_run_on_user_code_event_loop
     async def call_destructor(self):
         """Explicitly call the `__del__` method of the user callable.
 
-        We set the del method to noop after successfully calling it so the
-        destructor is called only once.
+        Calling this multiple times has no effect; only the first call will actually
+        call the destructor.
         """
-        async with self._delete_lock:
-            if not hasattr(self, "_callable"):
-                return
-
-            try:
-                if hasattr(self._callable, "__del__"):
-                    # Make sure to accept `async def __del__(self)` as well.
-                    await self._call_func_or_gen(self._callable.__del__)
-                    setattr(self._callable, "__del__", lambda _: None)
-
-                if hasattr(self._callable, "__serve_multiplex_wrapper"):
-                    await getattr(
-                        self._callable, "__serve_multiplex_wrapper"
-                    ).shutdown()
-
-            except Exception as e:
-                logger.exception(f"Exception during graceful shutdown of replica: {e}")
-            finally:
-                if hasattr(self._callable, "__del__"):
-                    del self._callable
+        if self._callable is None:
+            raise RuntimeError("initialize_callable must be called before destructor.")
+
+        # Only run the destructor once. This is safe because there is no `await` between
+        # checking the flag here and flipping it to `True` below.
+        if self._destructor_called:
+            return
```

Comment:
This logic to only run the destructor once was not actually working properly (discovered via unit testing). I simplified with a boolean guard.

Reply:
You don't need a separate boolean tho, you can just remove `self._callable` using it as idempotency token

Reply:
I did this originally but that interferes with the check to assert that the callable has been initialized. Having a separate boolean is pretty clear anyways 🤷